rasterio/_base.pyx

```diff
@@ -379,14 +379,13 @@ cdef class DatasetReader(object):
 
     def index(self, x, y):
         """Returns the (row, col) index of the pixel containing (x, y)."""
-        a, b, c, d, e, f, _, _, _ = self.affine
-        return int(math.floor((y-f)/e)), int(math.floor((x-c)/a))
+        return get_index(x, y, self.affine)
 
     def window(self, left, bottom, right, top, boundless=False):
         """Returns the window corresponding to the world bounding box.
         If boundless is False, window is limited to extent of this dataset."""
-        EPS = 1.0e-8
-        window = tuple(zip(self.index(left + EPS, top - EPS), self.index(right + EPS, bottom - EPS)))
+
+        window = get_window(left, bottom, right, top, self.affine)
```

Comment:
I'm :-1: on these changes in _base.pyx. The previous code is well tested – and less code.